src/number_theory/padics/padic_norm.lean

```diff
@@ -52,131 +52,153 @@ open_locale rat
-  { refl }
+  have neq_one : (¬ p = 1) ↔ true,
+  { exact iff_of_true ((ne_of_lt hp).symm) trivial, },
+  have eq_zero_false : (p = 0) ↔ false,
+  { exact iff_false_intro ((ne_of_lt (trans zero_lt_one hp)).symm) },
+  simp [padic_val_nat, neq_one, eq_zero_false],
 end
 
+end padic_val_nat
+
 /--
-`padic_val_rat p 0` is 0 for any `p`.
+For `p ≠ 1`, the p-adic valuation of an integer `z ≠ 0` is the largest natural number `k` such that
+p^k divides z.
+
+If `x = 0` or `p = 1`, then `padic_val_int p q` defaults to 0.
 -/
+def padic_val_int (p : ℕ) (z : ℤ) : ℕ :=
+padic_val_nat p (z.nat_abs)
+
+namespace padic_val_int
+open multiplicity
+variables {p : ℕ}
+
+/-- `padic_val_int p 0` is 0 for any `p`. -/
 @[simp]
-protected lemma zero (m : nat) : padic_val_rat m 0 = 0 := rfl
+protected lemma zero : padic_val_int p 0 = 0 :=
+by simp [padic_val_int]
 
-/--
-`padic_val_rat p 1` is 0 for any `p`.
--/
-@[simp] protected lemma one : padic_val_rat p 1 = 0 :=
-by unfold padic_val_rat; split_ifs; simp *
+/-- `padic_val_int p 1` is 0 for any `p`. -/
+@[simp] protected lemma one : padic_val_int p 1 = 0 :=
+by simp [padic_val_int]
+
+/-- For `p ≠ 0, p ≠ 1, `padic_val_rat p p` is 1. -/
+@[simp] lemma self (hp : 1 < p) : padic_val_int p p = 1 :=
+by simp [padic_val_int, padic_val_nat.self hp]
+
+end padic_val_int
 
 /--
-For `p ≠ 0, p ≠ 1, `padic_val_rat p p` is 1.
+`padic_val_rat` defines the valuation of a rational `q` to be the valuation of `q.num` minus the
+valuation of `q.denom`.
+If `q = 0` or `p = 1`, then `padic_val_rat p q` defaults to 0.
 -/
+def padic_val_rat (p : ℕ) (q : ℚ) : ℤ :=
+padic_val_int p q.num - padic_val_nat p q.denom
+
+namespace padic_val_rat
+open multiplicity
+variables {p : ℕ}
+
+/-- `padic_val_rat p q` is symmetric in `q`. -/
+@[simp] protected lemma neg (q : ℚ) : padic_val_rat p (-q) = padic_val_rat p q :=
+by simp [padic_val_rat, padic_val_int]
+
+/-- `padic_val_rat p 0` is 0 for any `p`. -/
+@[simp]
+protected lemma zero (m : nat) : padic_val_rat m 0 = 0 := by simp [padic_val_rat, padic_val_int]
+
+/-- `padic_val_rat p 1` is 0 for any `p`. -/
+@[simp] protected lemma one : padic_val_rat p 1 = 0 := by simp [padic_val_rat, padic_val_int]
+
+/-- For `p ≠ 0, p ≠ 1, `padic_val_rat p p` is 1. -/
 @[simp] lemma padic_val_rat_self (hp : 1 < p) : padic_val_rat p p = 1 :=
-by unfold padic_val_rat; split_ifs; simp [*, nat.one_lt_iff_ne_zero_and_ne_one] at *
+begin
+  simp only [padic_val_rat, padic_val_int, rat.coe_nat_num, int.nat_abs_of_nat, rat.coe_nat_denom,
+    padic_val_nat.one, int.coe_nat_zero, sub_zero, padic_val_nat.self hp],
+  refl,
+end
 
-/--
-The p-adic value of an integer `z ≠ 0` is the multiplicity of `p` in `z`.
--/
+/-- The p-adic value of an integer `z ≠ 0` is the multiplicity of `p` in `z`. -/
 lemma padic_val_rat_of_int (z : ℤ) (hp : p ≠ 1) (hz : z ≠ 0) :
   padic_val_rat p (z : ℚ) = (multiplicity (p : ℤ) z).get
     (finite_int_iff.2 ⟨hp, hz⟩) :=
-by rw [padic_val_rat, dif_pos]; simp *; refl
+begin
```

Comment:
This proof could then be golfed, if you prove the analogous lemma for `padic_val_int` first.

Reply:
Ok, I have golfed this proof a little bit with `padic_val_int.defn`